src/highdicom/seg/sop.py

```diff
@@ -1916,81 +1984,140 @@ def _get_pixels_by_seg_frame(
+                pixel_array = self.pixel_array
 
-            # Check for overlap by summing over the segments dimension
-            if np.any(pixel_array.sum(axis=-1) > 1):
-                raise RuntimeError(
-                    'Segments cannot be combined because they overlap'
-                )
+            if pixel_array.ndim == 2:
+                pixel_array = pixel_array[None, :, :]
 
-            # Scale the array by the segment number using broadcasting
-            if relabel:
-                pixel_value_map = np.arange(1, len(segment_numbers) + 1)
-            else:
-                pixel_value_map = segment_numbers
-            scaled_array = pixel_array * pixel_value_map.reshape(1, 1, 1, -1)
+            # Initialize empty pixel array
+            out_array = np.zeros((nfo, h, w), dtype=intermediate_dtype)
 
-            # Combine segments by taking maximum down final dimension
-            max_array = scaled_array.max(axis=-1)
-            pixel_array = max_array
+            # Loop over output frames
+            for fo in range(nfo):
+                # Loop over segmentation indices
+                for seg_ind, seg_n in enumerate(segment_numbers):
+                    fi = seg_frames_matrix[fo, seg_ind] - 1
+                    pix_value = intermediate_dtype(
+                        seg_ind + 1 if relabel else seg_n
+                    )
+                    if fi >= 0:
+                        if seg_ind > 0 and not skip_overlap_checks:
+                            if np.any(
+                                np.logical_and(
+                                    pixel_array[fi, :, :] > 0,
+                                    out_array[fo, :, :] > 0
+                                )
+                            ):
+                                raise RuntimeError(
+                                    "Cannot combine segments because segments "
+                                    "overlap."
+                                )
+                        out_array[fo, :, :] = np.maximum(
+                            pixel_array[fi, :, :] * pix_value,
+                            out_array[fo, :, :]
+                        )
 
-        return pixel_array
+        else:
+            # Initialize empty pixel array
+            out_array = np.zeros((nfo, h, w, nseg), intermediate_dtype)
+
+            # Loop through output frames
+            for out_frm, frm_row in enumerate(seg_frames_matrix):
+                # Loop through segmentation frames
+                for out_seg, seg_frame_num in enumerate(frm_row):
+                    if seg_frame_num >= 1:
+                        seg_frame_ind = seg_frame_num.item() - 1
+                        # Copy data to to output array
+                        if self.pixel_array.ndim == 2:
+                            # Special case with a single segmentation frame
+                            out_array[out_frm, :, :, out_seg] = \
+                                self.pixel_array
```

Comment:
Can you elaborate on your concern, not sure I understand?

Reply:
My concern is that NumPy may do a shallow copy and that `out_array` may continue to point to the array stored at `self.pixel_array` when it is returned by the function. With the references persisting, Python may not free the memory of `self.pixel_array` (even if the instance of the class is no longer needed).

It can sometimes be beneficial in my experience to enforce a deep copy for arrays that are returned from functions/methods.

https://numpy.org/doc/stable/user/quickstart.html#view-or-shallow-copy

Reply:
I'm almost certain that his isn't true after playing around with np.shared_memory but I added a .copy() just to be safe. Although the slicing operation could create a shallow copy, I'm pretty sure when you then assign this shallow copy as a value to a slice of the output array, the data is deepcopied. Otherwise the output array could consist of a bunch of different slices of other arrays, which is not allowed